src/ImageRunModal.jsx

```diff
@@ -27,13 +28,107 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports.splice(idx, 1);
+            if (ports.length === 0)
+                ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
```

Comment:
wow, 31337 way of updating the state :grin: Simply passing in the updated object would do as well, but we can get more familiar with the functional style.

But why does this need to copy the state in the onChange call?

Reply:
Well technically it could do without, but it propagates this change to the dialog itself, which then stores it in it's own state. And if the dialog would modifiy this part of the state, wouldn't it propagate to the state here?

Reply:
Ack, so let's keep this for safety.